lcls-twincat-math/lcls_twincat_math/POUs/FB_Quaternion.TcPOU

```diff
@@ -0,0 +1,376 @@
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[
+w := w + iQuat.w;
+x := x + iQuat.x;
+y := y + iQuat.y;
+z := z + iQuat.z;
+]]></ST>
+      </Implementation>
+    </Method>
+    <Method Name="DivQuat" Id="{1915c45d-0c4c-4cb5-9348-b641b28e4fe4}">
+      <Declaration><![CDATA[
+METHOD DivQuat
+VAR_INPUT
+    iQuat	: I_Quat;
+END_VAR
+VAR_OUTPUT
+    bError : BOOL;
+    sErrorMsg : T_MaxString;
+END_VAR
+VAR
+    fNormSq : LREAL;
+    fwi : LREAL;
+    fxi : LREAL;
+    fyi : LREAL;
+    fzi : LREAL;
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[
+bError := FALSE;
+sErrorMsg := '';
+
+fNormSq := F_NormSqQuat(iQuat);
+
+IF fNormSq = 0.0 THEN
+    bError := TRUE;
+    sErrorMsg := 'Error: cannot divide by a quaternion with a norm of 0.';
+    RETURN;
+END_IF
+
+fwi := w;
+fxi := x;
+fyi := y;
+fzi := z;
+
+w := ((iQuat.w * fwi) + (iQuat.x * fxi) + (iQuat.y * fyi) + (iQuat.z * fzi)) / fNormSq;
+x := ((iQuat.w * fxi) - (iQuat.x * fwi) - (iQuat.y * fzi) + (iQuat.z * fyi)) / fNormSq;
+y := ((iQuat.w * fyi) + (iQuat.x * fzi) - (iQuat.y * fwi) - (iQuat.z * fxi)) / fNormSq;
+z := ((iQuat.w * fzi) - (iQuat.x * fyi) + (iQuat.y * fxi) - (iQuat.z * fwi)) / fNormSq;
+]]></ST>
+      </Implementation>
+    </Method>
+    <Method Name="DivScalar" Id="{4d84e2f2-81a3-414a-b1e8-9a34fadfe7d1}">
+      <Declaration><![CDATA[
+METHOD DivScalar
+VAR_INPUT
+    fScalar : LREAL;
+END_VAR
+VAR_OUTPUT
+    bError : BOOL;
+    sErrorMsg : T_MaxString;
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[
+bError := FALSE;
+sErrorMsg := '';
+
+IF fScalar = 0.0 THEN
+    bError := TRUE;
+    sErrorMsg := 'Error: attempted division of quaternion by scalar of 0.';
+    RETURN;
+END_IF
+
+w := w / fScalar;
+x := x / fScalar;
+y := y / fScalar;
+z := z / fScalar;
+]]></ST>
+      </Implementation>
+    </Method>
+    <Method Name="EqualsQuat" Id="{b2ebfb21-208a-4535-9c1d-8fb4ad4e160f}">
+      <Declaration><![CDATA[
+METHOD EqualsQuat : BOOL
+VAR_INPUT
+    iQuat	: I_Quat;
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[
+EqualsQuat := w = iQuat.w AND
+              x = iQuat.x AND
+              y = iQuat.y AND
+              z = iQuat.z;
+]]></ST>
+      </Implementation>
+    </Method>
+    <Method Name="GetNorm" Id="{172c3d10-6a52-4e90-853d-ef52c393f88d}">
+      <Declaration><![CDATA[
+METHOD GetNorm : LREAL
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[
+GetNorm := SQRT(w*w + x*x + y*y + z*z);
+]]></ST>
+      </Implementation>
+    </Method>
+    <Method Name="GetNormSq" Id="{f4b27ab1-ff46-4a03-8e7e-5b8182f21d26}">
+      <Declaration><![CDATA[
+METHOD GetNormSq : LREAL
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[
+GetNormSq := w*w + x*x + y*y + z*z;
+]]></ST>
+      </Implementation>
+    </Method>
+    <Method Name="MakeConj" Id="{0d0a6a03-d7f4-464e-b9c6-f4816551c4f1}">
+      <Declaration><![CDATA[
+METHOD MakeConj
```

Comment:
I wonder if you have any thoughts about to what extent these sorts of methods need documentation? I know what these all mean but maybe someone else might not. But also, surely you'd only be here if you understand?